File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    answer = True
    dic={}
    flag=False
    for i in phone_book:
        dic[i]=len(i)
    #하나의 번호에 대해 다른 번호를 비교하는게 아니라(=최대 O(n^2)) 번호의 접두어가 dic에 존재하는지로 시간복잡도 줄임(=최대 O(n*20))
    for i in phone_book: 
        string="" #문자 하나씩 받아 접두어를 구성해 dic에 존재하는지 확인할 임의의 문자열 선언
        for j in i:
            string+=j #하나씩 받기
            if string in dic and string!=i: #하나씩 받은 접두어가 dic에 존재하고 현재 문자열이 아니면
                answer=False
                flag=True
                break
        if flag:
            break
    return answer

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;
class Solution {
    public boolean solution(String[] phone_book) {
        boolean answer = true;
        
        Map<String, Integer> map = new HashMap<>();
        
        for (String p: phone_book) {
            map.put(p, map.getOrDefault(p,0)+1);    
        }
        
        
        for (String p : phone_book) {
            String temp = "";
            for (char c: p.toCharArray()) {
                temp += c;
                if (map.containsKey(temp) && !temp.equals(p)) {
                    return false;
                }
            }
            
        }
        return true;
    }
}